✨ feat: 태그 추가/검색 기능 작성

Jiwoong0322 opened the pull request:
2. 변경 또는 추가사항을 아래에 간략하게 작성하세요 ( 필수 )

1 - 모집 공고 태그 생성 API 추가 (POST /api/v1/recruitments/tags/)
2 - 태그 목록 조회 및 자동 완성 검색 기능 구현 (GET /api/v1/recruitments/tags/)
3 - 페이지네이션(Paginator) 추가
4 - 태그 API 관련 테스트 코드 작성

3. 이슈가 있다면 아래에 연결하세요 ( 선택 ) 관련 이슈: #43

## ✅ PR 요약
- 관련 이슈 번호: #43 
- 작업 요약: 태그 추가/검색 기능 작성

## 📄 상세 내용
- [ ] 프로젝트에서 사용될 페이지네이션을 통합 관리할 paginators.py를 apps/core에 추가
- [ ] 태그 추가/검색 API 기능 작성 및 테스트 성공

## 📸 스크린샷 (선택)

## 📝 기타 참고 사항

## 🧪 PR Checklist
- [ ] 커밋 메시지 컨벤션에 맞게 작성했습니다.
- [ ] 변경 사항에 대한 테스트를 했습니다.(버그 수정/기능에 대한 테스트).

Jiwoong0322 committed:
✨ feat: 태그 추가/검색 기능 작성

2. 변경 또는 추가사항을 아래에 간략하게 작성하세요 ( 필수 )

1 - 모집 공고 태그 생성 API 추가 (POST /api/v1/recruitments/tags/)
2 - 태그 목록 조회 및 자동 완성 검색 기능 구현 (GET /api/v1/recruitments/tags/)
3 - 페이지네이션(Paginator) 추가
4 - 태그 API 관련 테스트 코드 작성

3. 이슈가 있다면 아래에 연결하세요 ( 선택 )
관련 이슈: #43

# Conflicts:
#	apps/recruitments/urls/recruitments_urls.py
#	config/urls.py

Umdoong commented on config/urls.py:
```diff
@@ -10,6 +10,7 @@
 urlpatterns: list[URLPattern | URLResolver] = [
     path("api/v1/auth/", include("apps.users.urls.auth_urls")),
     path("api/v1/notifications", include("apps.notifications.urls")),
+    path("api/v1/recruitments/", include("apps.recruitments.urls")),
```
api 명세서를 보니까 api/v1/recruitments로 get요청을 보내는 곳이 있네요! 끝에 /를 빼주세용!

Meoyoug commented on apps/core/paginators.py:
```diff
@@ -0,0 +1,13 @@
+
+# REQ-RECM-002/008 (태그 검색)에서 사용할 ㅍ페이지네이션
+# 페이지 당 5개의 항목을 반환하는 PageNumberPagination
+class FivePageNumberPagination(PageNumberPagination):
```
page_size를 고정된 값이 아니라 유동적이게 받을 수 있도록하면 필요한 곳들마다 유용하게 사용할 수 있지않을까요~?

Jiwoong0322 replied:
요구사항 정의서에는 페이지 당 항목 5개라고 지정이 되어있어서. page_size_query_param를 없앴는데 page_size가 유동적이게 받을 수 있게 되면 요구사항에 맞지 않게되지 않나요?

Meoyoug replied:
아뇨 지웅님 요지를 잘못파악하신 것 같은데 core에 정의한 것들은 모든 앱에서 공통으로 사용될 수 있는 모델이나 유틸, 서비스를 정의하는 곳인데 지금 지웅님이 작성하신 페이지네이션은 지웅님 쪽 api의 요구사항에 따라서만 page_size를 고정시켜뒀잖아요
page_size를 클래스를 인스턴스화 할때 유동적으로 받을 수 있게하면 API마다 필요한 페이지네이션 사이즈를 직접입력해서 페이지 네이션 클래스를 사용할 수 있다는 말입니다.

Jiwoong0322 replied:
아 그런거였군요. 제가 공식문서 글을 잘못 해석했네요. 수정하겠습니다

Meoyoug commented on apps/recruitments/serializers/tags_serializers.py:
```diff
@@ -0,0 +1,17 @@
+        fields = ["id", "name"]  # id와 name 필드를 클라이언트에게 반환하고, 데이터 유효성 검사를 수행한다.
+        # unique=True에 의해 자동으로 추가되는 UniqueValidator를 비활성화한다.
+        # 중복 검사는 Service Layer에서 직접 처리할것 .
+        extra_kwargs: ClassVar[Dict[str, Dict[str, List[Any]]]] = {
```
혹시 모델에서 지정된 밸리데이터를 제거한 이유가 있을까요?

Jiwoong0322 replied:
테스트 코드를 작성하고 테스트를 실행할 때 409 Conflict가 반환되어야 하는데 400 Bad Reqiest가 반환되어 문제를 찾다가 밸리데이터를 비활성화하면 해결될 수 있다는 걸 보고 밸리데이터를 비활성화 한 뒤 테스트를 진행하니 409 Conflict가 반환되어 테스트가 성공했기에 제거한채로 커밋을 진행했습니다.
하지만 리뷰를 보고 더욱 자세히 찾아보니 DRF의 기본 핸들러가 ValidationError가 발생했을 때 그 오류가 UniqueValidator때문인지, 아니면 다른 커스텀 유효성 검사때문인지를 자동 세분화 하여 판단하지 않는다는 것을 알고 좀 더 자세히 알아보는 중입니다.

Meoyoug replied:
말씀하신대로 DRF 의 모델 시리얼라이저를 사용하는 경우에 unique인 필드의 경우 UniqueValidator에 의해 이미 중복된 이름을 가진 태그가 db에 존재하는 지 확인하게 됩니다. 그래서 테스트코드에서도 400을 주는 것 같구요 에러메시지도 보시면 아마 중복된 이름을 가진 데이터가 존재한다고 나올 거에요
View나 서비스로직에서 409코드를 주는 부분이 중복이되는 셈이라 빼주시고 test에서 400으로 응답이 오는지, 응답메시지가 시리얼라이저에서 UniqueValidator에 의해 주는 에러메시지가 맞는지를 확인하는게 더 맞을 것 같습니다.

Meoyoug commented on apps/recruitments/services/tags_services.py:
```diff
@@ -0,0 +1,27 @@
+    # 새로운 태그 생성
+    def create_tag(self, name: str) -> Tag:
+
+        # 대소문자를 구분하지 않고 동일한 이름의 태그가 이미 존재하는지 확인한다.
```
이거 시리얼라이저에서 밸리데이터 제거안했으면 수행가능했을거라 한번 시도해보세용

Meoyoug commented on apps/recruitments/tests/tests.py:
```diff
@@ -1,3 +1,93 @@
+
+        # 테스트용 태그 6개 생성
+        self.tags_to_create = ["python", "django", "react", "javascript", "fastapi", "docker"]
+        for tag_name in self.tags_to_create:
```
여기 for 문으로 돌리면서 create를 호출하게되면 저 태그를 생성하는데 6번의 쿼리가 호출되어야하죠?
이거 1번으로 줄일 수 있는 방법이 있습니다!
한번 고민해보시면 좋을 것 같아요

Meoyoug commented on apps/recruitments/urls/recruitments_urls.py:
단순히 태그 기능이라면 파일명을 tag_urls.py 로 해주셔야할 것 같아요 나중에 공고관련 url을 작성하시는 분이 있을 텐데 urlpatterns가 너무 비대해져서 보기 힘들 것 같아요

Meoyoug commented on apps/recruitments/views/tags_views.py:
```diff
@@ -0,0 +1,61 @@
+from apps.recruitments.services.tags_services import TagService
+
+
+class TagViewSet(viewsets.ModelViewSet[Tag]):
```
이건 공고 작성 및 수정시에 사용되는 태그 검색 및 생성 뷰셋일까요?
공고 목록 조회시에도 태그를 사용한 필터링 적용이 있어서 권한만 다른 목록조회 기능이 필요하거든요

Umdoong replied:
ViewSet처럼 간단하게 구현할 수 있는 기능을 써보는 것도 좋지만, 정확하게 알고 쓰는 게 아니라면 APIView로 직접 구현해 보는 것이 더 좋을 것 같습니다.

Jiwoong0322 committed:
🎨 style: black, isort 코드 포매터 실행

2. 변경 또는 추가사항을 아래에 간략하게 작성하세요 ( 필수 )

isort, black 진행, apps/recruitments/tests/tests.py이름을 tests_recruitments_tags.py로 변경, config/urls.py의 api/v1/recruitments/ 끝에 /제거

3. 이슈가 있다면 아래에 연결하세요 ( 선택 )
관련 이슈: #43

Jiwoong0322 committed:
♻️ refactor:  ViewSet을 APIView로 전환 및 관련 코드 최적화`

2. 변경 또는 추가사항을 아래에 간략하게 작성하세요 ( 필수 )

1. PageNumberPagination 클래스를 클라이언트가 유동적으로 page_size를 변경할수있게 수정
2. recruitments_urls.py를 tags_urls.py로 이름을 바꾸고 관련 파일들 수정
3. tags_serializers.py의 Validator제거 코드 삭제.
4. tags_services.py의 409에러 발생 코드 제거. services에서 단순 생성 및 반환만 가능하도록 함.
5. tags_views.py를 ViewSet에서 APIView로 변경.
6. tests_recruitments_tags.py의 태그 생성 실패 테스트(이름중복) 개선.
7. isort, black실행, mypy, test통과.

3. 이슈가 있다면 아래에 연결하세요 ( 선택 )

관련 이슈: #54

Umdoong commented on apps/recruitments/tests/tests.py:
이 파일 삭제해주세용!

Jiwoong0322 replied:
이 파일 지우고 커밋한건데 안지워진건가요?

Meoyoug replied:
@Jiwoong0322 삭제된것맞습니다!

Umdoong commented on apps/recruitments/urls/tags_urls.py:
```diff
@@ -0,0 +1,9 @@
+urlpatterns = [
+    # GET, POST 요청 모두 동일한 URL을 사용하므로 하나의 path로 정의한다.
+    # .as_view() 메서드는 클래스 기반 뷰를 URL에 연결할 때 사용하는 표준 방식이다.
+    path("/tags/", TagAPIView.as_view(), name="tag-list"),
```
맨 끝에 / 뺴주세요

Umdoong commented on apps/recruitments/urls/__init__.py:
```diff
@@ -0,0 +1,7 @@
+
+app_name = "recruitments"
+
+urlpatterns = [
```
tags_urls.py말고 여기에 tag를 넣는 게 낫지 않을까용??

Umdoong commented on apps/recruitments/tests/tests_recruitments_tags.py:
```diff
@@ -0,0 +1,124 @@
+    # 1. 신규 태그 생성 테스트
+    def test_create_tag_success(self) -> None:
+        """태그 생성 성공 테스트"""
+        url = "/api/v1/recruitments/tags/"
```
하드코딩 말고 reverse를 활용하보시면 더 좋을 것 같아요!

Umdoong commented on apps/recruitments/views/tags_views.py:
```diff
@@ -0,0 +1,81 @@
+        # 2. 클래스 변수 'filter_backends'에 정의된 필터들을 순회하며 적용한다.
+        #    - SearchFilter가 'search_fields'를 참조하여, ?search=... 쿼리 파라미터가 있으면 쿼리셋을 필터링한다.
+        for backend in self.filter_backends:
+            queryset = backend().filter_queryset(request, queryset, self)
```
orm은 django.db.Manager를 통해 따로 함수로 만들어서 관리하면 view가 더 깔끔해질 것 같습니다.

Meoyoug reviewed (changes requested):
효종님이 코멘트 달아주신대로 __init__.py의 urlpatterns에서 prefix지정해주시고 푸시해주시면 머지할게요~

Jiwoong0322 committed:
♻️ refactor: url수정 및 테스트코드 수정

2. 변경 또는 추가사항을 아래에 간략하게 작성하세요 ( 필수 )

tags_urls.py의 /tags를 __init__.py에 적용
테스트 코드에 하드코딩 되어있던 url을 reverse를 사용하도록 변경

3. 이슈가 있다면 아래에 연결하세요 ( 선택 )

관련 이슈: #54

Jiwoong0322 committed:
💡 chore: recruitments urls 분리

recruitments default url, recruitments tag url 분리

Jiwoong0322 committed:
💡 chore: recruitments.urls 오타 수정

apps.recruitments.recruitments.urls -> apps.recruitments.urls.recruitments.urls

Jiwoong0322 committed:
♻️ refactor: recruitments urls 구조 변경

recruitments urls __init__.py import 형식으로 변경

Jiwoong0322 committed:
🎨 style: black, isort 코드 포매터 실행

isort, black, mypy 실행

관련 이슈: #54

Meoyoug reviewed (approved):
고생하셨어요 지웅님~ 머지할게용
